google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/AssignerBuilder.java

```diff
@@ -0,0 +1,98 @@
+ */
+
+package com.google.cloud.pubsublite.internal.wire;
+
+import com.google.auto.value.AutoValue;
+import com.google.cloud.pubsublite.Endpoints;
+import com.google.cloud.pubsublite.Stubs;
+import com.google.cloud.pubsublite.SubscriptionPath;
+import com.google.cloud.pubsublite.SubscriptionPaths;
+import com.google.cloud.pubsublite.proto.InitialPartitionAssignmentRequest;
+import com.google.cloud.pubsublite.proto.PartitionAssignmentServiceGrpc;
+import com.google.cloud.pubsublite.proto.PartitionAssignmentServiceGrpc.PartitionAssignmentServiceStub;
+import com.google.protobuf.ByteString;
+import io.grpc.Status;
+import io.grpc.StatusException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import java.util.UUID;
+
+@AutoValue
+public abstract class AssignerBuilder {
+  // Required parameters.
+  abstract SubscriptionPath subscriptionPath();
+
+  abstract PartitionAssignmentReceiver receiver();
+
+  // Optional parameters.
+  abstract Optional<PartitionAssignmentServiceStub> assignmentStub();
+
+  public static Builder newBuilder() {
+    return new AutoValue_AssignerBuilder.Builder();
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+    // Required parameters.
+    public abstract Builder setSubscriptionPath(SubscriptionPath path);
+
+    public abstract Builder setReceiver(PartitionAssignmentReceiver receiver);
+
+    // Optional parameters.
+    public abstract Builder setAssignmentStub(PartitionAssignmentServiceStub stub);
+
+    abstract AssignerBuilder autoBuild();
+
+    @SuppressWarnings("CheckReturnValue")
+    public Assigner build() throws StatusException {
+      AssignerBuilder builder = autoBuild();
+      SubscriptionPaths.check(builder.subscriptionPath());
+
+      PartitionAssignmentServiceStub stub;
+      if (builder.assignmentStub().isPresent()) {
+        stub = builder.assignmentStub().get();
+      } else {
+        try {
+          stub =
+              Stubs.defaultStub(
+                  Endpoints.regionalEndpoint(
+                      SubscriptionPaths.getZone(builder.subscriptionPath()).region()),
+                  PartitionAssignmentServiceGrpc::newStub);
+        } catch (IOException e) {
+          throw Status.INTERNAL
+              .withCause(e)
+              .withDescription("Creating cursor stub failed.")
+              .asException();
+        }
+      }
+
+      UUID uuid = UUID.randomUUID();
+      ByteBuffer uuidBuffer = ByteBuffer.allocate(16);
+      uuidBuffer.putLong(uuid.getMostSignificantBits());
+      uuidBuffer.putLong(uuid.getLeastSignificantBits());
+
+      InitialPartitionAssignmentRequest initial =
+          InitialPartitionAssignmentRequest.newBuilder()
+              .setSubscription(builder.subscriptionPath().value())
+              .setClientId(ByteString.copyFrom(uuidBuffer))
+              .build();
+      return new AssignerImpl(
+          stub, new ConnectedAssignerImpl.Factory(), initial, builder.receiver());
```

Comment:
Nit: check that receiver is actually populated?

Reply:
Autovalue does this already. its not optional.

Reply:
Autovalue does this already. Its not optional.